src/main/java/persistence/entity/CustomJpaRepository.java

```diff
@@ -0,0 +1,33 @@
+package persistence.entity;
+
+import persistence.sql.meta.Table;
+
+public class CustomJpaRepository<T, ID> implements Repository<T, ID> {
+
+    private final EntityManager entityManager;
+
+    public CustomJpaRepository(EntityManager entityManager) {
+        this.entityManager = entityManager;
+    }
+
+    @Override
+    public T save(T entity) {
+        if (isNew(entity)) {
+            return entityManager.persist(entity);
+        }
+        return entityManager.merge(entity);
+    }
+
+    private boolean isNew(T entity) {
+        Table table = Table.getInstance(entity.getClass());
+        Object id = table.getIdValue(entity);
+        if (id == null) {
+            return true;
+        }
+        else if (id instanceof Number) {
+            return ((Number) id).longValue() == 0L;
+        } else {
+            throw new IllegalArgumentException(String.format("Unsupported id type %s", id.getClass()));
+        }
```

Comment:
```suggestion
        if (id instanceof Number) {
            return ((Number) id).longValue() == 0L;
        }
        throw new IllegalArgumentException(String.format("Unsupported id type %s", id.getClass()));
```

(반영하지 않으셔도 됩니다.) 객체지향 생활 체조 원칙의 else 예약어를 쓰지 않는다라는 관점에서는 이렇게 표현되기도 해요.